### PR TITLE
gh-123049: Allow to create the unnamed section from scratch.

### LINKED ISSUE #123049: configparser: unable to create unnamed section
Hi, just chiming in because I really love this new feature - however, I haven't managed to create a new INI file from scratch with an unnamed section yet. Is this possible? Here are my futile attempts with configparser 7.0.0 from backports:

```
Python 3.12.4 (tags/v3.12.4:8e8a4ba, Jun  6 2024, 19:30:16) [MSC v.1940 64 bit (AMD64)] on win32
>>> from backports.configparser import UNNAMED_SECTION, ConfigParser
>>> cp = ConfigParser(allow_unnamed_section=True)
>>> cp.set(UNNAMED_SECTION, "option1", "hello world")
Traceback (most recent call last):
    super().set(section, option, value)
    raise NoSectionError(section) from None
backports.configparser.NoSectionError: No section: <UNNAMED_SECTION>
>>> cp.add_section(UNNAMED_SECTION)
  File "C:\Users\kerling\AppData\Local\pypoetry\Cache\virtualenvs\non-package-mode-l42eA81v-py3.12\Lib\site-packages\backports\configparser\__init__.py", line 1294, in _validate_value_types
    raise TypeError("section names must be strings")
TypeError: section names must be strings
>>> cp[UNNAMED_SECTION]["option1"] = "hello world"
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\kerling\AppData\Local\pypoetry\Cache\virtualenvs\non-package-mode-l42eA81v-py3.12\Lib\site-packages\backports\configparser\__init__.py", line 1070, in __getitem__
    raise KeyError(key)
KeyError: <UNNAMED_SECTION>
>>> cp[UNNAMED_SECTION] = { "option1": "hello world" }
>>> cp.sections()
['<UNNAMED_SECTION>']
>>> import io
>>> fp = io.StringIO()
>>> cp.write(fp)
>>> fp.getvalue()
'[<UNNAMED_SECTION>]\noption1 = hello world\n\n'
```

_Originally posted by @mtnpke in https://github.com/python/cpython/issues/66449#issuecomment-2290906624_
            

### Linked PRs
* gh-123077

### PR DESCRIPTION
**Pull request by pslacerda**
May fix #123049.

* Issue: gh-123049

### CODE REVIEW

File: Lib/configparser.py [hunk 1]
```diff
@@ -694,6 +694,7 @@ def add_section(self, section):
 
         if section in self._sections:
             raise DuplicateSectionError(section)
+
```
Comment (picnixz) [inline Lib/configparser.py]: Please revert unrelated cosmetic changes.

File: Misc/NEWS.d/next/Library/2024-08-16-16-53-52.gh-issue-123049.izx_fH.rst [hunk 2]
```diff
@@ -0,0 +1 @@
+Add support for the unnamed section in add_section.
```
Comment (picnixz) [inline Misc/NEWS.d/next/Library/2024-08-16-16-53-52.gh-issue-123049.izx_fH.rst]: This should be more precise, e.g.:

```rst
Add support for unnamed sections to
:meth:`ConfigParser.add_section <configparser.RawConfigParser.add_section>`.
```

I think this *should* work (but check that the class names are correct, maybe we can even just use `ConfigParser.add_section <configparser.ConfigParser.add_section>` instead, I'm not sure)

Comment (pslacerda) [reply Misc/NEWS.d/next/Library/2024-08-16-16-53-52.gh-issue-123049.izx_fH.rst]: Thanks...

**Comment by pslacerda** [comment]
I think now is fine.

File: Misc/NEWS.d/next/Library/2024-08-16-16-53-52.gh-issue-123049.izx_fH.rst (hunk 2 above)
Comment (jaraco) [reply Misc/NEWS.d/next/Library/2024-08-16-16-53-52.gh-issue-123049.izx_fH.rst]: I can see the docs tests failing due to a broken reference. You can figure out the best reference by looking at [the docs](https://docs.python.org/3/library/configparser.html#configparser.ConfigParser.add_section). Notice that the anchor is `#configparser.ConfigParser.add_section`. So what you need for the blurb is this:

```
Add support for unnamed sections to :meth:`configparser.ConfigParser.add_section`.
```

You can let sphinx find that anchor and determine how to display it.

File: Misc/NEWS.d/next/Library/2024-08-16-16-53-52.gh-issue-123049.izx_fH.rst [hunk 3]
```diff
@@ -0,0 +1 @@
+Add support for the unnamed section in :meth:`RawConfigParser.add_section`.
```
Comment (jaraco) [inline Misc/NEWS.d/next/Library/2024-08-16-16-53-52.gh-issue-123049.izx_fH.rst]:
```suggestion
Add support for the unnamed section in :meth:`configparser.ConfigParser.add_section`.
```

File: Lib/configparser.py [hunk 4]
```diff
@@ -1203,20 +1207,21 @@ def _convert_to_boolean(self, value):
         return self.BOOLEAN_STATES[value.lower()]
 
     def _validate_value_types(self, *, section="", option="", value=""):
-        """Raises a TypeError for non-string values.
+        """Raises a TypeError for incorrect non-string values.
```
Comment (jaraco) [inline Lib/configparser.py]: Maybe use the term 'illegal' values, since the next paragraph describes what's legal.

```suggestion
        """Raises a TypeError for illegal values.
```

**Review by jaraco — APPROVED** [review]
Looks good. Nice work including the test.

File: Misc/NEWS.d/next/Library/2024-08-16-16-53-52.gh-issue-123049.izx_fH.rst (hunk 3 above)
Comment (jaraco) [reply Misc/NEWS.d/next/Library/2024-08-16-16-53-52.gh-issue-123049.izx_fH.rst]: Not sure why that suggestion showed up incorrectly. I've corrected it.

**Review by jaraco — CHANGES_REQUESTED** [review]
Oh. I just noticed something else. By looking at the docs, I see they no longer match the implementation. We'll want to update the docs to reflect the implementation and also indicate the version changed where UNNAMED_SECTION is allowed (3.14).

**Comment by pslacerda** [comment]
Is this ok?

```rst
   .. method:: add_section(section)

      Add a section named *section* or :const:`UNNAMED_SECTION` to the instance.

      If the given section already exists, :exc:`DuplicateSectionError` is
      raised. If :const:`UNNAMED_SECTION` (with `allow_unnamed_section=True` on
      :meth:`!__init__`), or if the *default section* name is passed,
      :exc:`ValueError` is raised.

      Type of *section* is not checked which lets users create non-string named
      sections.  This behaviour is unsupported and may cause internal errors.

   .. versionchanged:: 3.14
        Added support for :const:`UNNAMED_SECTION`.
```

**Comment by pslacerda** [comment]
And the `_validate_value_types` check str explicitly, shoudn't it test for `__str__`?

```py
    def _validate_value_types(self, *, section="", option="", value=""):
        """Raises a TypeError for incorrect non-string values.

        Legal non-string values are UNNAMED_SECTION and valueless values if
        they are are allowed.

        For compatibility reasons this method is not used in classic set()
        for RawConfigParsers. It is invoked in every case for mapping protocol
        access and in ConfigParser.set().
        """
        if section is UNNAMED_SECTION:
            if not self._allow_unnamed_section:
                raise ValueError("UNNAMED_SECTION is not allowed")
        else:
            if not isinstance(section, str):
                raise TypeError("section names must be strings or UNNAMED_SECTION")
        if not isinstance(option, str):
            raise TypeError("option keys must be strings")
        if not self._allow_no_value or value:
            if not isinstance(value, str):
                raise TypeError("option values must be strings")
```

File: Lib/configparser.py [hunk 5]
```diff
@@ -692,6 +692,10 @@ def add_section(self, section):
         if section == self.default_section:
             raise ValueError('Invalid section name: %r' % section)
 
+        if section is UNNAMED_SECTION:
+            if not self._allow_unnamed_section:
+                raise ValueError("Unnamed section not enabled")
```
Comment (picnixz) [inline Lib/configparser.py]:
```suggestion
        if section is UNNAMED_SECTION and not self._allow_unnamed_section:
	   		raise ValueError("Support for UNNAMED_SECTION is disabled")
```

Do you perhaps want add a specific exception?

File: Lib/configparser.py [hunk 6]
```diff
@@ -1203,20 +1207,21 @@ def _convert_to_boolean(self, value):
...
 
         For compatibility reasons this method is not used in classic set()
         for RawConfigParsers. It is invoked in every case for mapping protocol
         access and in ConfigParser.set().
         """
-        if not isinstance(section, str):
-            raise TypeError("section names must be strings")
+        if section is UNNAMED_SECTION:
+            if not self._allow_unnamed_section:
+                raise ValueError("UNNAMED_SECTION is not allowed")
```
Comment (picnixz) [inline Lib/configparser.py]:
```suggestion
        if section is UNNAMED_SECTION and not self._allow_unnamed_section:
            raise ValueError("UNNAMED_SECTION is not allowed")
```

File: Lib/configparser.py [hunk 7]
```diff
@@ -1203,20 +1207,21 @@ def _convert_to_boolean(self, value):
...
+        else:
+            if not isinstance(section, str):
+                raise TypeError("section names must be strings or UNNAMED_SECTION")
```
Comment (picnixz) [inline Lib/configparser.py]:
```suggestion
        elif not isinstance(section, str):
             raise TypeError("section names must be strings or UNNAMED_SECTION")
```

File: Misc/NEWS.d/next/Library/2024-08-16-16-53-52.gh-issue-123049.izx_fH.rst [hunk 8]
```diff
@@ -0,0 +1 @@
+Add support for the unnamed section in :meth:`configparser.ConfigParser.add_section`.
```
Comment (picnixz) [inline Misc/NEWS.d/next/Library/2024-08-16-16-53-52.gh-issue-123049.izx_fH.rst]:
```suggestion
Add support for :const:`~configparser.UNNAMED_SECTION`
in :meth:`configparser.ConfigParser.add_section`.
```

File: Lib/configparser.py [hunk 9]
```diff
@@ -1203,20 +1207,21 @@ def _convert_to_boolean(self, value):
...
-        """Raises a TypeError for non-string values.
+        """Raises a TypeError for illegal values.
 
-        The only legal non-string value if we allow valueless
-        options is None, so we need to check if the value is a
-        string if:
-        - we do not allow valueless options, or
-        - we allow valueless options but the value is not None
+        Legal non-string values are UNNAMED_SECTION and valueless values if
+        they are are allowed.
```
Comment (picnixz) [inline Lib/configparser.py]:
```suggestion
        they are allowed.
```

File: Lib/configparser.py (hunk 5 above)
Comment (pslacerda) [reply Lib/configparser.py]: Yes, just added.

File: Doc/library/configparser.rst [hunk 10]
```diff
@@ -1314,13 +1314,19 @@ RawConfigParser Objects
...
+      If the given section already exists, :exc:`DuplicateSectionError` is
+      raised. If the *default section* name is passed, :exc:`ValueError` is
+      raised. If :const:`UNNAMED_SECTION` is passed and support is disabled,
+      :exc:`UnnamedSectionDisabledError` is raised.
 
       Type of *section* is not checked which lets users create non-string named
       sections.  This behaviour is unsupported and may cause internal errors.
 
+   .. versionchanged:: 3.14
+        Added support for :const:`UNNAMED_SECTION`.
```
Comment (picnixz) [inline Doc/library/configparser.rst]:
```suggestion
   .. versionchanged:: 3.14
      Added support for :const:`UNNAMED_SECTION`.
```

File: Lib/configparser.py [hunk 11]
```diff
@@ -692,6 +700,10 @@ def add_section(self, section):
         if section == self.default_section:
             raise ValueError('Invalid section name: %r' % section)
 
+        if section is UNNAMED_SECTION:
+            if not self._allow_unnamed_section:
+                raise UnnamedSectionDisabledError()
```
Comment (picnixz) [inline Lib/configparser.py]: I suggested a new exception because there was the `DuplicateSectionError` that existed, so I wasn't sure whether another exception would have been relevant.

File: Doc/library/configparser.rst [hunk 12]
```diff
@@ -1314,13 +1314,19 @@ RawConfigParser Objects
...
 
-      Add a section named *section* to the instance.  If a section by the given
-      name already exists, :exc:`DuplicateSectionError` is raised.  If the
-      *default section* name is passed, :exc:`ValueError` is raised.
+      Add a section named *section* or :const:`UNNAMED_SECTION` to the instance.
+
+      If the given section already exists, :exc:`DuplicateSectionError` is
+      raised. If the *default section* name is passed, :exc:`ValueError` is
+      raised. If :const:`UNNAMED_SECTION` is passed and support is disabled,
+      :exc:`UnnamedSectionDisabledError` is raised.
```
Comment (picnixz) [inline Doc/library/configparser.rst:1322]: You need to document this new exception.

File: Lib/configparser.py (hunk 11 above)
Comment (pslacerda) [reply Lib/configparser.py]: Yes, I agree that this was the best option.

File: Doc/library/configparser.rst (hunk 12 above)
Comment (pslacerda) [reply Doc/library/configparser.rst:1322]: Check the new commit, I'm not sphinx testing... only when pushing.

File: Doc/library/configparser.rst [hunk 13]
```diff
@@ -1421,6 +1426,13 @@ Exceptions
 
    .. versionadded:: 3.13
 
+.. exception:: UnnamedSectionDisabledError
+
+   Exception raised when attempting to use the unnamed section without enabling
+   it.
```
Comment (picnixz) [inline Doc/library/configparser.rst]:
```suggestion
   Exception raised when attempting to use the
   :const:`UNNAMED_SECTION` without enabling it.
```

File: Doc/library/configparser.rst [hunk 14]
```diff
@@ -1403,8 +1409,7 @@ Exceptions
 .. exception:: MissingSectionHeaderError
 
    Exception raised when attempting to parse a file which has no section
-   headers.
-
+   headers..
```
Comment (picnixz) [inline Doc/library/configparser.rst]:
```suggestion
   headers.
```

File: Lib/configparser.py (hunk 11 above)
Comment (picnixz) [inline Lib/configparser.py]: You could omit parentheses when raising an exception without argument, but I'm not sure if this is what we usually do (so maybe check the rest of the module / rest of the library).

File: Doc/library/configparser.rst (hunk 13 above)
Comment (pslacerda) [reply Doc/library/configparser.rst]: I didn't explicitly talked about `UNNAMED_SECTION` because it doesn't need to be explicitly refereed when parsing a file, except for allowing it with `enable_unnamed_section=True`.

**Review by jaraco — APPROVED** [review]
Looks great. Thanks for the extended effort!

File: Doc/library/configparser.rst (hunk 13 above)
Comment (jaraco) [reply Doc/library/configparser.rst]: I'll await resolution of this concern before merging. @picnixz let us know if pslacerda's explanation satisfies your concern.

**Comment by pslacerda** [comment]
Thanks @jaraco and @picnixz.

Please, squash the messy commits on merging.

File: Doc/library/configparser.rst (hunk 13 above)
Comment (picnixz) [reply Doc/library/configparser.rst]: I am fine with that. What matters to me is that people understand why such exception could be raised in general. I would have mentioned the constant so that you have a clickable link but it's also ok without it (it could make the docs slightly more readable without the link).

**Review by picnixz — APPROVED** [review]

**Comment by jaraco** [comment]
Great work everyone. I've backported this functionality to [configparser](https://pypi.org/project/configparser) 7.1, so you can utilize the behavior from there until you can rely on Python 3.14.

**Comment by mtnpke** [comment]
Works perfectly! Thanks!
